kratos/geometries/quadrilateral_3d_4.h

```diff
@@ -1524,6 +1525,83 @@ template<class TPointType> class Quadrilateral3D4
         return rResult;
     }
 
+    ///@}
+    ///@name Spatial Operations
+    ///@{
+
+    /**
+    * @brief Projects a certain point on the geometry, or finds
+    *        the closest point, depending on the provided
+    *        initial guess. The external point does not necessary
+    *        lay on the geometry.
+    *        It shall deal as the interface to the mathematical
+    *        projection function e.g. the Newton-Raphson.
+    *        Thus, the breaking criteria does not necessarily mean
+    *        that it found a point on the surface, if it is really
+    *        the closest if or not. It shows only if the breaking
+    *        criteria, defined by the tolerance is reached.
+    *
+    *        This function requires an initial guess, provided by
+    *        rProjectedPointLocalCoordinates.
+    *        This function can be a very costly operation.
+    *
+    * @param rPointGlobalCoordinates the point to which the
+    *        projection has to be found.
+    * @param rProjectedPointGlobalCoordinates the location of the
+    *        projection in global coordinates.
+    * @param rProjectedPointLocalCoordinates the location of the
+    *        projection in local coordinates.
+    *        The variable is as initial guess!
+    * @param Tolerance accepted of orthogonal error to projection.
+    * @return It is chosen to take an int as output parameter to
+    *         keep more possibilities within the interface.
+    *         0 -> failed
+    *         1 -> converged
+    */
+    int ProjectionPoint(
+        const CoordinatesArrayType& rPointGlobalCoordinates,
+        CoordinatesArrayType& rProjectedPointGlobalCoordinates,
+        CoordinatesArrayType& rProjectedPointLocalCoordinates,
+        const double Tolerance = std::numeric_limits<double>::epsilon()
+        ) const override
+    {
+        // Max number of iterations
+        const std::size_t max_number_of_iterations = 100;
+
+        // We do a first guess in the center of the geometry
+        auto projected_point = this->Center();
+        array_1d<double, 3> normal = this->UnitNormal(projected_point);
+
+        // SOme auxiliar variables
+        const Point point_to_project(rPointGlobalCoordinates);
+        double distance;
+        std::size_t iter;
+
+        // We iterate until we find the properly projected point
+        for (iter = 0; iter < max_number_of_iterations; ++iter) {
+            // We compute the distance, if it is not in the plane we project
+            projected_point = GeometricalProjectionUtilities::FastProject( projected_point, point_to_project, normal, distance);
+
+            // If the normal corresponds means that we have converged
+            if (norm_2(this->UnitNormal(projected_point) - normal) < Tolerance * 1.0e4) break;

```

Comment:
what if it is flat? the normal would be always the same...so you would detect convergence anyway.

and what if the closest point is on the edge, and the projection is not along the normal?

Reply:
looks a bit unusual at first sight.

Please check if `std::numeric_limits<double>::epsilon()` is a good default value for the tolerance. IMHO it is quite arbitrary, far too small and leads to the strange `Tolerance * 1.0e4`.

ad 2: Usually I project to the edges in a first step. There might be a faster approach but it is simple and reliable.